spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/AbstractStringBasedJpaQuery.java

```diff
@@ -179,4 +188,77 @@ protected String potentiallyRewriteQuery(String originalQuery, Sort sort, @Nulla
 				? queryRewriter.rewrite(originalQuery, pageable) //
 				: queryRewriter.rewrite(originalQuery, sort);
 	}
+
+	String applySorting(CachableQuery cachableQuery) {
+
+		return QueryEnhancerFactory.forQuery(cachableQuery.getDeclaredQuery()).applySorting(cachableQuery.getSort(),
+				cachableQuery.getAlias());
+	}
+
+	private String applySortingIfNecessary(DeclaredQuery query, Sort sort) {
+
+		if (sort.isUnsorted()) {
+			return query.getQueryString();
+		}
+		return applySorting(query, sort);
+	}
+
+	/**
+	 * Value object with optimized {@link Object#equals(Object)} to cache a query based on its query string and
+	 * {@link Sort sorting}.
+	 *
+	 * @since 3.2.3
+	 * @author Christoph Strobl
+	 */
+	static class CachableQuery {
```

Comment:
The caching key is already quite complex. I'd suggest to keep only the `isUnsorted` code path and refine the way we render queries with a sort. That is, render everything before and after `ORDER BY` in a first pass (e.g. held in a `QueryEnhancer` instance). Then, render `Sort` with rather simple string concatenation.

Reply:
The above is deliberately targeting the 3.2.x line without applying major changes.

Reply:
Created #3326 to follow up on splitting the transformation (targeting _main_ development line)